martinize2/processors/canonize_PTMs.py

```diff
@@ -0,0 +1,268 @@
+                return node1['atomname'] == node2['atomname']
+        else:
+            return False
+
+
+def find_PTM_atoms(molecule):
+    """
+    Finds all atoms in molecule that have the node attribute ``PTM_atom`` set
+    to a value that evaluates to ``True``. ``molecule`` will be traversed
+    starting at these atoms untill all marked atoms are visited such that they
+    are identified per "branch", and for every branch the anchor node is known.
+    The anchor node is the node(s) which are not PTM atoms and share an edge
+    with the traversed branch.
+
+    Parameters
+    ----------
+    molecule : networkx.Graph
+
+    Returns
+    -------
+    list of tuples of two sets of node indices
+        ``[({ptm atom indices}, {anchor indices}), ...]``. Ptm atom indices are
+        connected, and are connected to the rest of molecule via anchor
+        indices.
+    """
+
+    # Atomnames have already been fixed, and missing atoms have been added.
+    # In addition, unrecognized atoms have been labeled with the PTM attribute.
+    extra_atoms = set(n_idx for n_idx in molecule
+                      if molecule.nodes[n_idx].get('PTM_atom', False))
+    PTMs = []
+    while extra_atoms:
+        # First PTM atom we'll look at
+        first = next(iter(extra_atoms))
+        anchors = set()
+        # PTM atoms we've found
+        atoms = set()
+        # Atoms we still need to see this traversal
+        to_see = set([first])
+        # Traverse in molecule.
+        for orig, succ in nx.bfs_successors(molecule, first):
+            # We've seen orig, so remove it
+            to_see.remove(orig)
+            if orig in extra_atoms:
+                # If this is a PTM atom, we want to see it's neighbours as
+                # well.
+                to_see.update(succ)
+                atoms.add(orig)
+            else:
+                # Else, it's an attachment point for the this PTM
+                anchors.add(orig)
+            if not to_see:
+                # We've traversed the interesting bit of the tree
+                break
+        extra_atoms -= atoms
+        PTMs.append((atoms, anchors))
+    return PTMs
+
+
+def identify_ptms(residue, residue_ptms, known_PTMs):
+    """
+    Identifies all PTMs in ``known_PTMs`` nescessary to describe all PTM atoms in
+    ``residue_ptms``. Will take PTMs such that all PTM atoms in ``residue``
+    will be covered by applying PTMs from ``known_PTMs`` in order.
+    Nodes in ``residue`` must have correct ``atomname`` attributes, and may not
+    be missing. In addition, every PTM in must be anchored to a non-PTM atom.
+
+    Parameters
+    ----------
+    residue : networkx.Graph
+        The residues involved with these PTMs. Need not be connected.
+
+    residue_ptms : list of tuples of two sets of node indices
+        As returned by ``find_PTM_atoms``, but only those relevant for
+        ``residue``.
+
+    known_PTMs : sequence of tuples of (networkx.Graph, PTMGraphMatcher)
```

Comment:
Is this not yet clear enough?

Reply:
I was expecting PTMs to gave more. Aren't them named? Do they only need/use the "atomname" attribute? There is at least the 'replace' attribute to document somewhere. An example would not hurt.

Reply:
Fair enough. Which attributes of the Graph are used is not really described. I'll get on it.

Reply:
Thanks. Now I can write PTM graphs for the tests.